src/libutil/file-system.cc

```diff
@@ -765,4 +765,19 @@ bool isExecutableFileAmbient(const fs::path & exe) {
         ) == 0;
 }
 
+namespace fs {
+
+std::filesystem::path parent_canonical(const std::filesystem::path & path)
+{
+    auto parent = path.parent_path();
+    if (parent == path) {
+        // `path` is a root directory => trivially canonical
+        return parent;
+    }
+    return std::filesystem::canonical(parent) / path.filename();
```

Comment:
Please wrap this into a NixError if this is throwing exceptions so we get nice error messages.